contrib/babelfishpg_tsql/src/procedures.c

```diff
@@ -2437,84 +2440,122 @@ update_bbf_server_options(char *servername, char *optname, char *optvalue, bool
-									 	RowExclusiveLock);
-		bbf_servers_def_rel_dsc = RelationGetDescr(bbf_servers_def_rel);
+			if(strlen(optname) == 13 && strncmp(optname, "query timeout", 13) == 0)
+			{
+				new_record_repl[Anum_bbf_servers_def_query_timeout - 1] = true;
+				new_record[Anum_bbf_servers_def_query_timeout - 1] = Int32GetDatum(timeout);
+			}
+			else
+			{
+				new_record_repl[Anum_bbf_servers_def_connect_timeout - 1] = true;
+				new_record[Anum_bbf_servers_def_connect_timeout - 1] = Int32GetDatum(timeout);
+			}
+		}
+		else
+		{
+			ereport(ERROR,
+				(errcode(ERRCODE_FDW_ERROR),
+				 errmsg("Invalid option provided for sp_serveroption")));
+		}
+	}
 
-		MemSet(new_record_nulls, false, sizeof(new_record_nulls));
-		MemSet(new_record_repl, false, sizeof(new_record_repl));
+	bbf_servers_def_rel = table_open(get_bbf_servers_def_oid(),RowExclusiveLock);
+	bbf_servers_def_rel_dsc = RelationGetDescr(bbf_servers_def_rel);
 
-		new_record[Anum_bbf_servers_def_servername - 1] = CStringGetTextDatum(servername);
-		new_record[Anum_bbf_servers_def_query_timeout - 1] = Int32GetDatum(query_timeout);
+	MemSet(new_record_nulls, false, sizeof(new_record_nulls));
+	new_record[Anum_bbf_servers_def_servername - 1] = CStringGetTextDatum(servername);
 
-		if(isInsert)
+	if(isInsert)
+	{
+		tuple = heap_form_tuple(bbf_servers_def_rel_dsc,
+								new_record, new_record_nulls);
+		CatalogTupleInsert(bbf_servers_def_rel, tuple);
+	}
+	else
+	{
+		ScanKeyInit(&key,
+					Anum_bbf_servers_def_servername,
+					BTEqualStrategyNumber, F_TEXTEQ,
+					CStringGetTextDatum(servername));
+		tblscan = table_beginscan_catalog(bbf_servers_def_rel, 1, &key);
+		old_tuple = heap_getnext(tblscan, ForwardScanDirection);
+
+		if (!old_tuple)
 		{
-			tuple = heap_form_tuple(bbf_servers_def_rel_dsc,
-									new_record, new_record_nulls);
-			CatalogTupleInsert(bbf_servers_def_rel, tuple);
+			table_endscan(tblscan);
+			table_close(bbf_servers_def_rel, RowExclusiveLock);
+			ereport(ERROR,
+					(errcode(ERRCODE_FDW_ERROR),
+					errmsg("The server '%s' does not exist. Use sp_linkedservers to show available servers.", servername)));
 		}
-		else
+
+		for(int i = 1; i < BBF_SERVERS_DEF_NUM_COLS; i++)
 		{
-			/* Search and obtain the tuple based on the server_id */	
-			ScanKeyInit(&key,
-						Anum_bbf_servers_def_servername,
-						BTEqualStrategyNumber, F_TEXTEQ,
-						CStringGetTextDatum(servername));
-			tblscan = table_beginscan_catalog(bbf_servers_def_rel, 1, &key);
-			old_tuple = heap_getnext(tblscan, ForwardScanDirection);
-
-			if (!old_tuple)
+			if(!new_record_repl[i])
 			{
-				table_endscan(tblscan);
-				table_close(bbf_servers_def_rel, RowExclusiveLock);
-				ereport(ERROR,
-						(errcode(ERRCODE_FDW_ERROR),
-				 		errmsg("The server '%s' does not exist. Use sp_linkedservers to show available servers.", servername)));
+				bool isNull;
+				new_record[i] = heap_getattr(old_tuple, i+1,
+												RelationGetDescr(bbf_servers_def_rel), &isNull);
```

Comment:
we need not fetch the attributes again, this adds unnecessary overhead. We can simply set the new_record[] like we are doing for the servername